nodejs-env/src/main/scala/org/scalajs/jsenv/nodejs/ComSupport.scala

```diff
@@ -0,0 +1,276 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Node.js env       **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2017, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.jsenv.nodejs
+
+import java.io._
+import java.net._
+
+import org.scalajs.core.tools.io.{VirtualJSFile, MemVirtualJSFile}
+import org.scalajs.jsenv._
+
+import scala.collection.immutable
+import scala.concurrent._
+import scala.util.control.NonFatal
+
+// TODO Replace this by a better execution context on the RunConfig.
+import scala.concurrent.ExecutionContext.Implicits.global
+
+private final class ComRun(run: JSRun, handleMessage: String => Unit,
+    serverSocket: ServerSocket) extends JSComRun {
+  import ComRun._
+
+  /** Promise that completes once the reciever thread is completed. */
+  private[this] val promise = Promise[Unit]()
+
+  @volatile
+  private[this] var state: State = AwaitingConnection(Nil)
+
+  // If the run fails, we make sure we also fail.
+  run.future.failed.foreach(forceClose _)
+
+  // TODO replace this with scheduled tasks on the execution context.
+  private[this] val receiver = new Thread {
+    setName("ComRun receiver")
+
+    override def run(): Unit = {
+      try {
+        awaitConnection()
+
+        while (!promise.isCompleted) {
```

Comment:
This is not the right condition to loop on. In the close procedure, it will loop until the async callback is called.